blockchain/blockindex.go

```diff
@@ -525,6 +733,76 @@ func (bi *blockIndex) removeChainTip(tip *blockNode) {
 	}
 }
 
+// forEachChainTip calls the provided function with each chain tip known to the
+// block index.  Returning an error from the provided function will stop the
+// iteration early and return said error from this function.
+//
+// This function MUST be called with the block index lock held (for reads).
+func (bi *blockIndex) forEachChainTip(f func(tip *blockNode) error) error {
+	for _, tipEntry := range bi.chainTips {
+		if err := f(tipEntry.tip); err != nil {
+			return err
+		}
+		for _, tip := range tipEntry.otherTips {
+			if err := f(tip); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// forEachChainTipAfterHeight calls the provided function with each chain tip
+// known to the block index that has a height which is greater than the provided
+// filter node.
+//
+// Providing a filter node also makes use of the recent chain tip cache when
+// possible which typically further reduces the number of chain tips that need
+// to be iterated since all old chain tips are pruned from the cache.
+//
+// Returning an error from the provided function will stop the iteration early
+// and return said error from this function.
+//
+// This function MUST be called with the block index lock held (for reads).
+func (bi *blockIndex) forEachChainTipAfterHeight(filter *blockNode, f func(tip *blockNode) error) error {
+	// Use the cached recent chain tips when the filter height permits it.
+	if filter.height >= bi.cachedTipsStart {
```

Comment:
It is zero by default and all chain tips are added to the cache when initially loading the index from storage and then it is immediately pruned, so, yes, it will always be correct.

Also, great observation on the inclusive starting height!  That is indeed the case.

Updated to use `filter.height >= bi.cachedTipsStart - 1`.